### ISSUE TITLE
Compatibility with React Compiler

### ISSUE THREAD

**Issue by sofi-manomano** [cvazac/eslint-plugin-react-perf#79]
I am currently using React Compiler in my app, which automatically handles the `useCallback`, `useMemo` and `memo` optimisations. However, the rule [react-perf/jsx-no-new-function-as-prop](https://github.com/cvazac/eslint-plugin-react-perf/blob/master/docs/rules/jsx-no-new-function-as-prop.md) is complaining when I am not wrapping with a `useCallback`, the functions declared within a component. So, for example, in the following component:

```
export const Accordion = ({ children, title }: AccordionProps) => {
  const [isCollapsed, setCollapsed] = useState(true);
  const toggle = useCallback(() => setCollapsed((prev) => !prev), []);  <--------------- HERE

  return (
    <View>
      <Pressable onPress={toggle} style={[styles.header, !isCollapsed && styles.headerOpen]}>
          <Text>{title}</Typography>
      </Pressable>

      <View style={[styles.content, isCollapsed && styles.collapsed]}>
        {children}
      </View>
    </View>
  );
};
```

I would like to have the rule detect this is wrong:

```
<Pressable onPress={() => setCollapsed((prev) => !prev)}>
```

And to make me declare the function outside the JSX, like this:

```
const toggle = () => setCollapsed((prev) => !prev);
<Pressable onPress={toggle}>
```

But not have to wrap it with the unnecessary useCallback, since I have React Compiler configurated. So my request is, could you add a custom config for the rule to avoid creating functions within the render, but not require the use of `useCallback`?

**Comment by hilja** [comment]
Some observations after playing in the React Compiler playground.

Inline function prop:

```js
const list = ['foo', 'bar', 'baz']
export default function MyApp() {
  const [foo, setFoo] = useState(false)

  return list.map(item => {
    console.log(foo)
    return <button onClick={() => setFoo(prev => !prev)}>{item}</button>
  })
}
```

Turns into [this](https://playground.react.dev/#N4Igzg9grgTgxgUxALhHCA7MAXABAGwEsdcBeXAbQHIAzCCKgGlyoCMBDGJljgLyoC6AHQwIAHgAcIMPABMENdlHx4aUDHGyFMuALIBPAIISJACgCUuYCNy50WPBToRmYBNgBi9AWVxQ3AMrY7NgIpor4buY2uDDusBgExNgAdAC27GaEoWlkAHxWMbb2kPgIKfgQAObh9NGJtrHxMIkAPKxQ2Ng6mADCRHAA1qTAFvm4bp70phJxAG7jAISzCHPmAL55wNkIaeutAPQdXZh5Mev16yIgjGiYNIRVKCCEaVIyuNj6EghWuAAK+CgVUIGAA8hItJgwLh1rgaDAILk2OxWAh8ABaCRAkEYDFxdiaDHoN6EMowA6yZJUADcIhEpmsDQOBxJEjJIW0GF0EHkyFwQhA7Hw+EFIjhYE5YAeCBhgOBoIhUKw5hpN3AAAsIAB3ACSGFCLWFYBQETc6yAA):

```js
import { c as _c } from "react/compiler-runtime";
const list = ["foo", "bar", "baz"];
export default function MyApp() {
  const $ = _c(2);
  const [foo, setFoo] = useState(false);
  let t0;
  if ($[0] !== foo) {
    t0 = list.map((item) => {
      console.log(foo);
      return <button onClick={() => setFoo(_temp)}>{item}</button>;
    });
    $[0] = foo;
    $[1] = t0;
  } else {
    t0 = $[1];
  }
  return t0;
}
function _temp(prev) {
  return !prev;
}
```

And the outside function prop:

```js
const list = ['foo', 'bar', 'baz']
export default function MyApp() {
  const [foo, setFoo] = useState(false)
  const toggle = () => setFoo(prev => !prev)

  return list.map(item => {
    // Comment this out to see a change
    console.log(foo)
    return <button onClick={toggle}>{item}</button>
  })
}
```

Turns into [this](https://playground.react.dev/#N4Igzg9grgTgxgUxALhHCA7MAXABAGwEsdcBeXAbQHIAzCCKgGlyoCMBDGJljgLyoC6AHQwIAHgAcIMPABMENdlHx4aUDHGyFMuALIBPAIISJACgCUuYCNy50WPBToRmYBNgBi9AWVxQ3AMrY7NgIpor4buY2dpgk2BAA5on4CL4WZAB8uG6e9KYSMAgAblm4AISFJdEYtkXYsLVEOAB0ALbsZoShbWXWtba4APRDuADCEG1tCBh42AAWxLjQcxA5CGnsdvPsGIkIMbb2kKkt+Enh9DWDuPWNuAA8rFDYCbWYY0RwANakwAnJVIAX0ywG6CDaQIeQ2er0wmRiQJqQJEIEYaEwNEIiRQIEIbSkMlw2H0EjSwFwAAV8FBEoQMAB5CRaOK4IG4GgwSY8disBD4AC0EhpdIwAqK7E0AvQBMIqRgQ1kxGwVAA3CIRKZ+rYRjKJHKQtoMLoIPJkLghCB2Ph8JaROywIawFiEGAqSL6UyWVhzKq0eB5hAAO4ASVmCBgGGtYBQETcQKAA) if you use `foo` inside the loop:

```js
import { c as _c } from "react/compiler-runtime";
const list = ["foo", "bar", "baz"];
export default function MyApp() {
  const $ = _c(3);
  const [foo, setFoo] = useState(false);
  let t0;
  if ($[0] === Symbol.for("react.memo_cache_sentinel")) {
    t0 = () => setFoo(_temp);
    $[0] = t0;
  } else {
    t0 = $[0];
  }
  const toggle = t0;
  let t1;
  if ($[1] !== foo) {
    t1 = list.map((item) => {
      console.log(foo);
      return <button onClick={toggle}>{item}</button>;
    });
    $[1] = foo;
    $[2] = t1;
  } else {
    t1 = $[2];
  }
  return t1;
}
function _temp(prev) {
  return !prev;
}
```

The outside `toggle` is a bit larger (about 34%) because it does two memoizations. But the `toggle` function is not invoked on every iteration.

- Outside `toggle` (minified): 852 bytes
- Inline `toggle` (minified): 605 bytes

Your app wont be 34% larger for sure, maybe few percentages max. Nothing to worry about.
